src/github.com/matrix-org/dendrite/mediaapi/writers/upload.go

```diff
@@ -0,0 +1,149 @@
+// limitations under the License.
+
+package writers
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/mediaapi/config"
+	"github.com/matrix-org/dendrite/mediaapi/types"
+	"github.com/matrix-org/util"
+)
+
+// uploadRequest metadata included in or derivable from an upload request
+// https://matrix.org/docs/spec/client_server/r0.2.0.html#post-matrix-media-r0-upload
+// NOTE: The members come from HTTP request metadata such as headers, query parameters or can be derived from such
+type uploadRequest struct {
+	MediaMetadata *types.MediaMetadata
+	Logger        *log.Entry
+}
+
+// uploadResponse defines the format of the JSON response
+// https://matrix.org/docs/spec/client_server/r0.2.0.html#post-matrix-media-r0-upload
+type uploadResponse struct {
+	ContentURI string `json:"content_uri"`
+}
+
+// Upload implements /upload
+// This endpoint involves uploading potentially significant amounts of data to the homeserver.
+// This implementation supports a configurable maximum file size limit in bytes. If a user tries to upload more than this, they will receive an error that their upload is too large.
+// Uploaded files are processed piece-wise to avoid DoS attacks which would starve the server of memory.
+// TODO: We should time out requests if they have not received any data within a configured timeout period.
+func Upload(req *http.Request, cfg *config.MediaAPI) util.JSONResponse {
+	r, resErr := parseAndValidateRequest(req, cfg)
+	if resErr != nil {
+		return *resErr
+	}
+
+	// doUpload
+
+	return util.JSONResponse{
+		Code: 200,
+		JSON: uploadResponse{
+			ContentURI: fmt.Sprintf("mxc://%s/%s", cfg.ServerName, r.MediaMetadata.MediaID),
+		},
+	}
+}
+
+// parseAndValidateRequest parses the incoming upload request to validate and extract
+// all the metadata about the media being uploaded.
+// Returns either an uploadRequest or an error formatted as a util.JSONResponse
+func parseAndValidateRequest(req *http.Request, cfg *config.MediaAPI) (*uploadRequest, *util.JSONResponse) {
+	if req.Method != "POST" {
+		return nil, &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.Unknown("HTTP request method must be POST."),
+		}
+	}
+
+	// authenticate user
+
+	r := &uploadRequest{
+		MediaMetadata: &types.MediaMetadata{
+			Origin:             cfg.ServerName,
+			ContentDisposition: types.ContentDisposition(req.Header.Get("Content-Disposition")),
+			FileSizeBytes:      types.FileSizeBytes(req.ContentLength),
+			ContentType:        types.ContentType(req.Header.Get("Content-Type")),
+			UploadName:         types.Filename(req.FormValue("filename")),
+		},
+		Logger: util.GetLogger(req.Context()),
+	}
+
+	if resErr := r.Validate(cfg.MaxFileSizeBytes); resErr != nil {
+		return nil, resErr
+	}
+
+	// FIXME: do we want to always override ContentDisposition here or only if
+	// there is no Content-Disposition header set?
```

Comment:
Synapse seems to have bailed out never implemented parsing the `Content-Disposition` header from the upload request. https://github.com/matrix-org/synapse/blob/fad3a8433535d7de321350bbd17373138c6fd3ec/synapse/rest/media/v1/upload_resource.py#L91

It instead takes the upload name from a `filename` query parameter: https://github.com/matrix-org/synapse/blob/fad3a8433535d7de321350bbd17373138c6fd3ec/synapse/rest/media/v1/upload_resource.py#L71

It parses them for remote files downloads but only stores the upload name, if specified: https://github.com/matrix-org/synapse/blob/fad3a8433535d7de321350bbd17373138c6fd3ec/synapse/rest/media/v1/media_repository.py#L169

If the upload name is present in the db, it responds to download/thumbnail requests with a Content-Disposition header including the file name: https://github.com/matrix-org/synapse/blob/9bba6ebaa903a81cd94fada114aa71e20b685adb/synapse/rest/media/v1/_base.py#L66

The override is essentially ignoring the `Content-Disposition` header from the request and placing higher priority on the `filename` query parameter.

Given synapse ignores the `Content-Disposition` header on uploads, the override would likely result in the same behaviour as synapse. But given the TODO status in synapse for parsing the header, it's not clear what is supposed to be done.